staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+)
+
+// MaxHashBits is the max bit length which can be used from hash value.
+// If we use all bits of hash value, the critical(last) card shuffled by
+// Dealer will be uneven to 2:3 (first half:second half) at most,
+// in order to reduce this unevenness to 32:33, we set MaxHashBits to 60 here.
+const MaxHashBits = 60
+
+// RequiredEntropyBits will figure out the required entropy bits from the given
+// deckSize and handSize by calculating bits of (deckSize^handSize).
+// RequiredEntropyBits must be less or equal than MaxHashBits
+// if we want to create and use Dealer.
+func RequiredEntropyBits(deckSize, handSize int) int {
+	return big.NewInt(1).Exp(big.NewInt(int64(deckSize)), big.NewInt(int64(handSize)), nil).BitLen()
```

Comment:
We do not need `big` here, and there is an off-by-1 error here.  Following is a much simpler and more efficient way to do this.
```
import "math/bits"
...
    return bits.Len64(deckSize-1)*handSize
```

Reply:
I agree we shouldn't need `big`, but I think RequiredEntropyBits should defend against overflow (e.g, handSize musn't be larger than MaxInt / 64 in Mike's version above - which is ridiculously large, arbitrarily choosing 100 or 50 as the limit is fine. Negative values should also result in failing - which in this case means returning MaxInt as the required entropy bits -- I think that's easiest.

Reply:
Oops, my suggestion has to be modified to get the types right.  It should be
```
import "math/bits"
...

func NewDealer(deckSize, handSize int) (*Dealer, error) {
...
        if deckSize > 1 << 26 {
                return nil, fmt.Errorf("deckSize %d is impractically large", deckSize)
        }
...
}
...


func RequiredEntropyBits(deckSize, handSize int) int {
    return bits.Len(uint(deckSize-1))*handSize
}
```

@lavalamp : do you realize that the other checks in the NewDealer constructor, together with the one I suggested adding, ensure that there is no overflow or confusion due to negative numbers?  Note that my suggested implementation of RequiredEntropyBits will not panic with any parameter values.  Note also that returning MaxInt is not exactly a clear way to say that the parameters were invalid; it is a hack that works in the intended context, which is the tests in the NewDealer constructor or any similar code.  As long as we are presuming the intended context, I think it is sufficient to document the parameter restrictions in a comment.  Following is a draft of such a comment.

```
// RequiredEntropyBits makes a conservative estimate of the number of bits of
// hash value that are consumed in shuffle sharding a deck of the given size to
// a hand of the given size.  The result is meaningful only if
// 1 <= handSize <= deckSize <= 1<<26.
```

Reply:
The 26th power of 2 is the highest one with the property that when multiplied by its base-2 logarithm the result is guaranteed to fit into a 32-bit signed integer without overflow.  https://golang.org/ref/spec#Numeric_types says that `int` is a 32 or 64 bit integer.

Reply:
If someone wants to argue that we must repeat the parameter checks inside RequiredEntropyBits --- making for three copies of those checks in our code --- then I think this is an argument that we are doing something wrong.  If we can not extract that function from its context without internalizing a copy of that context then, well, I think that is an argument that this function simply can not be extracted from its context.  If that is to be argued then I think the best course of action is to eliminate the RequiredEntropyBits function and put the required expression directly in the checking code in the constructor of Dealer.

Reply:
@lavalamp @MikeSpreitzer  I don't think we have reached a solution about where should we put and how should we use the validations in shuffle sharding package... 
I agree on the limited max value for deckSize from @lavalamp , 1<<26 seems too large for validation, maybe 1024 or 512 is more proper here?
And I don't think the off-by-1 error will stop us from using `bigint`, because bitsof(deckSize-1)*handSize is less accurate as bitsof(deckSize^handSize-1),  https://play.golang.org/p/5C9rB_0RxMT ,  in some cases, the difference between the two algorithms is 7 (36-29), I think this is unacceptable.

Reply:
Oh, right.  I was focusing only on deckSize values that are a power of two.  The estimation formula I suggested can indeed err in other cases.

There _is_ indeed an off-by-one error in the PR's current formulation using `math/big`.

See https://gist.github.com/MikeSpreitzer/206d449cb23e74fe8486895e483865ff for a comparison of three techniques for extimatimg the number of entropy bits:
```
big.NewInt(1).Sub(big.NewInt(1).Exp(big.NewInt(int64(deckSize)), big.NewInt(int64(handSize)), nil), big.NewInt(1)).BitLen()
```
,
```
int(math.Ceil( math.Log2(float64(deckSize)) * float64(handSize) ))
```
, and
```
int(math.Ceil( math.Log2(float64(deckSize)) * float64(handSize) - 1e-11))
```

It shows that all three produce identical results for 2 <= deckSize <= 1<<30 and result <= 60.

I did a little investigation of whether we could rely on `math.Log2` of a power of 2 producing an exactly correct answer.  I suspect that maybe we can.  Section 4.3 of IEEE Std 754-2008 (both as of 2008 and in the 2019 revision) says

> Except where stated otherwise, every operation shall be performed as if it first produced an intermediate result correct to infinite precision and with unbounded range, and then rounded that result according to one of the attributes in this clause.

The attributes control how rounding is done, but rounding is moot when the infinite precision result is an integer.

When deckSize is not a power of 2, we need not worry about the fact that floating point numbers are in general approximations because the error of the approximation is certain to be too small to make a difference in the particular cases we care about.

Thus, I favor the simplest code, to wit:
```
import "math"
...

// RequiredEntropyBits makes a quick and slightly conservative estimate of the number
// of bits of hash value that are consumed in shuffle sharding a deck of the given size
// to a hand of the given size.  The result is meaningful only if
// 1 <= handSize <= deckSize .
func RequiredEntropyBits(deckSize, handSize int) int {
    return int(math.Ceil( math.Log2(float64(deckSize)) * float64(handSize) ))
}
```

Reply:
I asked the Sophers math channel on Slack; the conversation _ended_ at https://gophers.slack.com/archives/CC84BTU6L/p1571278799066200 .  The answer is that Go's `math` package does _not_ generally "round correctly" according to IEEE Std 754-2008 and 2019.  However, we got lucky: we do not care about generalities, all we really need to know is whether `math.Log2` of `1<<k` produces exactly `k`.  And it does!  The source code is at https://golang.org/src/math/log10.go

So this confirms to me that my latest suggestion can be expected to work.

Reply:
I tried in https://play.golang.org/p/cO4_6D5XtKf and got the same result as @MikeSpreitzer , I will abandon math/big ...

Reply:
Updated as your suggestion.